File: src/core/spec/feature-tree-spec-resource-contract.ts
import { existsSync, readFileSync } from "fs";
import path from "path";
import { fileURLToPath } from "url";

import featureTreeSpecManifestJson from "../../../resources/specialists/specs/feature-tree/manifest.json";

export type FeatureTreeFrameworkSpecEntry = {
  id: string;
  title: string;
  resourceUri: string;
  fileName: string;
  description: string;
  signals: string[];
};

export type FeatureTreeSpecManifest = {
  schemaVersion: number;
  id: string;
  description: string;
  baseRulesInPrompt: boolean;
  availableSpecIds: string[];
  selectionRules: string[];
  specs: FeatureTreeFrameworkSpecEntry[];
};

export type FeatureTreeSpecResolvedResource = {
  uri: string;
  mimeType: "application/json" | "text/markdown";
  text: string;
};

export const FEATURE_TREE_SPEC_MANIFEST_RESOURCE_URI =
  "resource://routa/specialists/feature-tree/manifest";

const featureTreeSpecManifest = featureTreeSpecManifestJson as FeatureTreeSpecManifest;
const featureTreeSpecModuleDir = path.dirname(fileURLToPath(import.meta.url));

function collectSpecialistRootCandidates(baseDir: string): string[] {
  const resolvedBaseDir = path.resolve(baseDir);
  const candidates: string[] = [];
  let current = resolvedBaseDir;

  while (true) {
    candidates.push(path.join(current, "specialists"));
    candidates.push(path.join(current, "resources", "specialists"));

    const parent = path.dirname(current);
    if (parent === current) {
      break;
    }
    current = parent;
  }

  return candidates;
}

function resolveFeatureTreeSpecRootDir(): string {
  const envRoot = process.env.ROUTA_SPECIALISTS_RESOURCE_DIR?.trim();
  const candidateRoots = [
    ...(envRoot ? [
      path.resolve(envRoot),
      path.resolve(envRoot, "specialists"),
      path.resolve(envRoot, "resources", "specialists"),
    ] : []),
    ...collectSpecialistRootCandidates(featureTreeSpecModuleDir),
  ];

  const uniqueRoots = [...new Set(candidateRoots)];
  for (const root of uniqueRoots) {
    const specRoot = path.join(root, "specs", "feature-tree");
    const manifestPath = path.join(specRoot, "manifest.json");
    if (!existsSync(manifestPath)) {
      continue;
    }

    const allSpecFilesExist = featureTreeSpecManifest.specs.every((spec) =>
      spec.fileName
      && path.basename(spec.fileName) === spec.fileName
      && existsSync(path.join(specRoot, spec.fileName)),
    );
    if (allSpecFilesExist) {
      return specRoot;
    }
  }

  throw new Error(
    "Unable to resolve bundled feature-tree specialist specs. " +
    "Set ROUTA_SPECIALISTS_RESOURCE_DIR or run from a traced Routa build.",
  );
}

const featureTreeSpecRootDir = resolveFeatureTreeSpecRootDir();

export function getFeatureTreeSpecManifest(): FeatureTreeSpecManifest {
  return featureTreeSpecManifest;
}

export function getFeatureTreeSpecResourceUris(): string[] {
  return getFeatureTreeSpecManifest().specs.map((spec) => spec.resourceUri);
}

export function getFeatureTreeSpecEntryById(id: string): FeatureTreeFrameworkSpecEntry | undefined {
  return getFeatureTreeSpecManifest().specs.find((spec) => spec.id === id);
}

export function readFeatureTreeSpecResource(uri: string): FeatureTreeSpecResolvedResource | null {
  if (uri === FEATURE_TREE_SPEC_MANIFEST_RESOURCE_URI) {
    return {
      uri,
      mimeType: "application/json",
      text: `${JSON.stringify(getFeatureTreeSpecManifest(), null, 2)}\n`,
    };
  }

  const spec = getFeatureTreeSpecManifest().specs.find((entry) => entry.resourceUri === uri);
  if (!spec) {
    return null;
  }

  const filePath = path.join(featureTreeSpecRootDir, spec.fileName);
  const text = readFileSync(filePath, "utf-8");

  return {
    uri,
    mimeType: "text/markdown",
    text,
  };
}

export function buildFeatureTreeSpecPromptSection(): string {
  const manifest = getFeatureTreeSpecManifest();
  const exampleUris = manifest.specs
    .map((spec) => `- ${spec.id}: ${spec.resourceUri}`)
    .join("\n");

  return [
    "Framework overlay spec access:",
    "- Base feature-grouping rules are already built into this system prompt.",
    `- Read the manifest first: \`${FEATURE_TREE_SPEC_MANIFEST_RESOURCE_URI}\`.`,
    "- If your provider supports MCP resources/read directly, use that.",
    "- Otherwise call MCP tool `read_specialist_spec_resource` with the same URI.",
    `- Current bundled framework spec ids: ${manifest.availableSpecIds.join(", ")}.`,
    "- After reading the manifest, load only the framework specs supported by repository evidence.",
    "- A repository may need multiple framework specs when it has multiple runtime surfaces, such as Next.js plus Axum.",
    "- If no matching framework spec exists, continue with the built-in path model and repository evidence only.",
    "- Available framework spec resources:",
    exampleUris,
  ].join("\n");
}

File: crates/routa-core/src/models/feature_tree_spec_resource_contract.rs
use serde::{Deserialize, Serialize};

pub const FEATURE_TREE_SPEC_MANIFEST_RESOURCE_URI: &str =
    "resource://routa/specialists/feature-tree/manifest";

const FEATURE_TREE_SPEC_MANIFEST_JSON: &str = include_str!(concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/../../resources/specialists/specs/feature-tree/manifest.json"
));
const FEATURE_TREE_SPEC_NEXTJS: &str = include_str!(concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/../../resources/specialists/specs/feature-tree/nextjs.spec.md"
));
const FEATURE_TREE_SPEC_AXUM: &str = include_str!(concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/../../resources/specialists/specs/feature-tree/axum.spec.md"
));
const FEATURE_TREE_SPEC_EXPRESS: &str = include_str!(concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/../../resources/specialists/specs/feature-tree/express.spec.md"
));
const FEATURE_TREE_SPEC_SPRING_BOOT: &str = include_str!(concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/../../resources/specialists/specs/feature-tree/spring-boot.spec.md"
));
const FEATURE_TREE_SPEC_GIN: &str = include_str!(concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/../../resources/specialists/specs/feature-tree/gin.spec.md"
));
const FEATURE_TREE_SPEC_EGGJS: &str = include_str!(concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/../../resources/specialists/specs/feature-tree/eggjs.spec.md"
));

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct FeatureTreeFrameworkSpecEntry {
    pub id: String,
    pub title: String,
    pub resource_uri: String,
    pub file_name: String,
    pub description: String,
    pub signals: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct FeatureTreeSpecManifest {
    pub schema_version: usize,
    pub id: String,
    pub description: String,
    pub base_rules_in_prompt: bool,
    pub available_spec_ids: Vec<String>,
    pub selection_rules: Vec<String>,
    pub specs: Vec<FeatureTreeFrameworkSpecEntry>,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct FeatureTreeSpecResolvedResource {
    pub uri: String,
    pub mime_type: String,
    pub text: String,
}

const FEATURE_TREE_SPEC_FILES: &[(&str, &str)] = &[
    ("nextjs", FEATURE_TREE_SPEC_NEXTJS),
    ("axum", FEATURE_TREE_SPEC_AXUM),
    ("express", FEATURE_TREE_SPEC_EXPRESS),
    ("spring-boot", FEATURE_TREE_SPEC_SPRING_BOOT),
    ("gin", FEATURE_TREE_SPEC_GIN),
    ("eggjs", FEATURE_TREE_SPEC_EGGJS),
];

pub fn get_feature_tree_spec_manifest() -> FeatureTreeSpecManifest {
    serde_json::from_str(FEATURE_TREE_SPEC_MANIFEST_JSON)
        .expect("feature-tree specialist spec manifest should parse")
}

pub fn get_feature_tree_spec_resource_uris() -> Vec<String> {
    get_feature_tree_spec_manifest()
        .specs
        .into_iter()
        .map(|spec| spec.resource_uri)
        .collect()
}

pub fn read_feature_tree_spec_resource(uri: &str) -> Option<FeatureTreeSpecResolvedResource> {
    if uri == FEATURE_TREE_SPEC_MANIFEST_RESOURCE_URI {
        return Some(FeatureTreeSpecResolvedResource {
            uri: uri.to_string(),
            mime_type: "application/json".to_string(),
            text: format!("{FEATURE_TREE_SPEC_MANIFEST_JSON}\n"),
        });
    }

    let manifest = get_feature_tree_spec_manifest();
    let spec = manifest
        .specs
        .iter()
        .find(|entry| entry.resource_uri == uri)?;
    FEATURE_TREE_SPEC_FILES
        .iter()
        .find(|(id, _)| *id == spec.id)
        .map(|(_, source)| FeatureTreeSpecResolvedResource {
            uri: uri.to_string(),
            mime_type: "text/markdown".to_string(),
            text: (*source).to_string(),
        })
}

pub fn build_feature_tree_spec_prompt_section() -> String {
    let manifest = get_feature_tree_spec_manifest();
    let example_uris = manifest
        .specs
        .iter()
        .map(|spec| format!("- {}: {}", spec.id, spec.resource_uri))
        .collect::<Vec<_>>()
        .join("\n");

    [
        "Framework overlay spec access:".to_string(),
        "- Base feature-grouping rules are already built into this system prompt.".to_string(),
        format!(
            "- Read the manifest first: `{FEATURE_TREE_SPEC_MANIFEST_RESOURCE_URI}`."
        ),
        "- If your provider supports MCP resources/read directly, use that.".to_string(),
        "- Otherwise call MCP tool `read_specialist_spec_resource` with the same URI.".to_string(),
        format!(
            "- Current bundled framework spec ids: {}.",
            manifest.available_spec_ids.join(", ")
        ),
        "- After reading the manifest, load only the framework specs supported by repository evidence.".to_string(),
        "- A repository may need multiple framework specs when it has multiple runtime surfaces, such as Next.js plus Axum.".to_string(),
        "- If no matching framework spec exists, continue with the built-in path model and repository evidence only.".to_string(),
        "- Available framework spec resources:".to_string(),
        example_uris,
    ]
    .join("\n")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn reads_feature_tree_spec_manifest_resource() {
        let resource = read_feature_tree_spec_resource(FEATURE_TREE_SPEC_MANIFEST_RESOURCE_URI)
            .expect("expected manifest resource");

        assert_eq!(resource.mime_type, "application/json");
        assert!(resource
            .text
            .contains("\"id\": \"feature-tree-framework-specs\""));
        assert!(resource.text.contains(
            "\"resourceUri\": \"resource://routa/specialists/feature-tree/specs/nextjs\""
        ));
    }

    #[test]
    fn reads_feature_tree_spec_resource() {
        let resource =
            read_feature_tree_spec_resource("resource://routa/specialists/feature-tree/specs/axum")
                .expect("expected axum resource");

        assert_eq!(resource.mime_type, "text/markdown");
        assert!(resource.text.contains("# Axum Feature Surface Overlay"));
        assert!(resource.text.contains("Router::new()"));
    }

    #[test]
    fn builds_feature_tree_prompt_section_with_manifest_uri() {
        let prompt = build_feature_tree_spec_prompt_section();

        assert!(prompt.contains("read_specialist_spec_resource"));
        assert!(prompt.contains(FEATURE_TREE_SPEC_MANIFEST_RESOURCE_URI));
        assert!(prompt.contains("nextjs"));
        assert!(prompt.contains("axum"));
    }
}
